envoy/http/filter.h

```diff
@@ -319,6 +319,9 @@ class DownstreamStreamFilterCallbacks {
    */
   virtual void
   requestRouteConfigUpdate(RouteConfigUpdatedCallbackSharedPtr route_config_updated_cb) PURE;
+
+  virtual uint32_t maxRequestHeadersKb() const PURE;
```

Comment:
Can we add this API to the HeaderMap interface? It looks like a better encapsulation than adding it to the DownstreamStreamFilterCallbacks

Reply:
+1

Reply:
Thanks for the comments!  

I had been trying to add this to the HeaderMap interface. But it feels like this needs some heavy lifting to get the HCM limits plumbed into all the places which construct HeaderMaps, like headers, trailers,downstream(decoder processing), upstream(encoder processing), also different codes: H1/H2/H3.    I don't see a center place to get this done easily. Please share your thoughts.

Reply:
I think as long as the header map we're actually interacting with in ext_proc comes from a place that has properly constructed it based on HCM limits, we're OK. Or at least that HCM has had the chance to set the limits on the header map on its way to ext_proc. So, we can by default have an unbounded header map, but the invariant should be that by time it's observed by ext_proc it has become a bounded header map.

Reply:
Thanks for sharing the thoughts and ideas!

Where plumbing these HCM limits into HeaderMap interface does seem to have better encapsulation. However, IMHO, plumbing these HCM properties into HeaderMap does have memory cost. ----- 16 bytes per header map. In that regards, the solution in the current PR is more efficient since basically it is a call into the HCM to get those numbers.

Reply:
I'll defer to @yanavlasov on this, more a suggestion on my part. I'd note that 16 bytes per header map is likely negligible for any header map representing a HTTP request, and we sholdn't be copying header maps frequently in the context of a request (or we have other problems..).

Reply:
I created this PR for the efforts to plumb max_header_count/ max_header_size limits to HeaderMap.
https://github.com/envoyproxy/envoy/pull/28131

The implementation sets up the MAX limits during header/trailer creation moment in H1 and H2 codecs based on an offline discussion with @yanavlasov.  I will come back and update this PR: #27525 after #28131 merges.